### PR TITLE
[Validate] GR106

### PR DESCRIPTION
**Pull request by shmuel44**
## Related Issues
fixes: [CIAC-11253](https://jira-dc.paloaltonetworks.com/browse/CIAC-11253)
content-docs: [PR](https://github.com/demisto/content-docs/pull/1583) link

## Description
New validate
Ensure that each test playbook is linked to at least one content item.

### CODE REVIEW

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 1]
```diff
@@ -566,6 +567,12 @@ def find_mandatory_hidden_packs_dependencies(
             self._add_relationships_to_objects(session, results)
             return [self._id_to_obj[result] for result in results]
 
+    def find_test_playbook_without_uses(
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:570]:
```suggestion
    def find_unused_test_playbook(
```

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 2]
```diff
@@ -308,3 +308,13 @@ def validate_duplicate_ids(
...
+
+
+def validate_test_playbook_in_use(tx: Transaction, test_playbook_id: str) -> bool:
+    query = f"""// Return the name of the Test Playbook if is no any content-item that use this Test Playbook
+MATCH (tp:TestPlaybook{{object_id: '{test_playbook_id}'}})
+WHERE (NOT EXISTS (()-[:TESTED_BY]->(tp)))
+AND  tp.deprecated =false
+RETURN collect(tp.name) AS tp_name
+    """
+    return bool([item.get("tp_name") for item in run_query(tx, query) if item.get("tp_name")])
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:320]:
```suggestion
    return any((item.get("tp_name") for item in run_query(tx, query))
```

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ValidationResult,
+)
+
+ContentTypes = TestPlaybook
+
+
+class IsTestPlaybookInUseValidator(BaseValidator[ContentTypes], ABC):
+    error_code = "GR106"
+    description = (
+        "Ensure that each test playbook is linked to at least one content item."
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py:18]:
```suggestion
        "Checks that every test playbook is linked to at least one content item. (the content item has a 'tests:' key with the id of the test playbook)"
```

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    rationale = (
+        "Proper linkage of test playbooks ensures comprehensive testing. For guidelines,"
+        " visit: https://xsoar.pan.dev/docs/integrations/test-playbooks#adding-the-playbook-to-your-project"
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py:22]:
```suggestion
        "In the demisto/content repo, unlinked test playbooks are not run in PRs unless the test playbook itself is modified. Proper linkage of test playbooks ensures content quality. "
        "See https://xsoar.pan.dev/docs/integrations/test-playbooks#adding-the-playbook-to-your-project"
```

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    error_message = "Test playbook '{}' is not linked to any content item. Please ensure it is properly utilized."
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py:24]:
```suggestion
    error_message = "Test playbook '{}' is not linked to any content item. Make sure at least one integration, script or playbook mention it under the `tests:` key."
```

File: demisto_sdk/commands/validate/tests/GR_validators_test.py [hunk 6]
```diff
@@ -227,3 +234,49 @@ def test_MarketplacesFieldValidatorAllFiles_is_valid(
...
+
+    Then:
+    - Ensure that the validator correctly identifies the playbook in use with no errors.
+    - Ensure that the validator correctly identifies the playbook not in use and returns an appropriate error message.
+    - Ensure that the validator correctly identifies the deprecated playbook with no errors.
+    """
+    graph_interface = prepared_graph_repo.create_graph()
+    BaseValidator.graph_interface = graph_interface
+    playbook_in_use = (
+        prepared_graph_repo.packs[1].test_playbooks[0].get_graph_object(graph_interface)
```
Comment (dorschw) [inline demisto_sdk/commands/validate/tests/GR_validators_test.py:258]:
```suggestion
    graph_interface = prepared_graph_repo.create_graph()
    BaseValidator.graph_interface = graph_interface
    pack = prepared_graph_repo.packs[1]
    playbook_in_use = (
        pack.test_playbooks[0].get_graph_object(graph_interface)
```
reuse the `pack` variable in lines 266, 277 for better readability

File: TestSuite/pack.py [hunk 7]
```diff
@@ -16,6 +16,7 @@
     XSIAM_DASHBOARDS_DIR,
     XSIAM_REPORTS_DIR,
 )
+from demisto_sdk.commands.content_graph.objects.test_playbook import TestPlaybook
```
Comment (dorschw) [inline TestSuite/pack.py:19]: a little suspicious that it's the only graph object imported here, let's make sure with @YuvHayun this is the right place.

**Review by dorschw — COMMENTED** [review]
Nice

File: TestSuite/pack.py (hunk 7 above)
Comment (shmuel44) [reply TestSuite/pack.py:19]: @YuvHayun WDYT?

Comment (shmuel44) [reply TestSuite/pack.py:19]: @YuvHayun approved

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 8]
```diff
@@ -566,6 +567,16 @@ def find_mandatory_hidden_packs_dependencies(
             self._add_relationships_to_objects(session, results)
             return [self._id_to_obj[result] for result in results]
 
+    def find_unused_test_playbook(
+        self, test_playbook_ids: List[str], skipped_tests_keys: list
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:571]:
```suggestion
        self, test_playbook_ids: List[str], skipped_tests_keys: List[str]
```
?

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 9]
```diff
@@ -1,5 +1,6 @@
 from typing import Dict, List, Tuple
 
+from more_itertools import first
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:3]:
```suggestion
from more_itertools import one
```

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 10]
```diff
@@ -308,3 +309,28 @@ def validate_duplicate_ids(
...
+AND NOT EXISTS {{ MATCH ()-[:TESTED_BY]->(tp) }}
+AND tp.deprecated = false
+AND NOT (tp.object_id IN {tests_skipped})
+MATCH (tp)-[:IN_PACK]->(p:Pack)
+WHERE p.support = "xsoar"
+AND p.deprecated = false
+RETURN collect(tp) AS content_items
+"""
+    # when there a test playbooks that not in use, the query return a list with one item
+    return first(
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:329]:
```suggestion
    return one(
```
this also fails if there's `0` or `>1` values

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class IsTestPlaybookInUseValidator(BaseValidator[ContentTypes], ABC):
+    error_code = "GR106"
+    description = (
+        "Checks that every test playbook is linked to at least one content item."
+        " (the content item has a 'tests:' key with the id of the test playbook)"
+    )
+    rationale = (
+        "In the demisto/content repo, unlinked test playbooks are not run in PRs unless the test playbook itself is modified. Proper linkage of test playbooks ensures content quality. "
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py:24]:
```suggestion
        "In the demisto/content repo, unlinked test playbooks are not run in CI (for PRs) unless the test playbook itself is modified. Proper linkage of test playbooks ensures content quality. "
```

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        "See  https://xsoar.pan.dev/docs/integrations/test-playbooks#adding-the-playbook-to-your-project"
+    )
+    error_message = "Test playbook '{}' is not linked to any content item. Make sure at least one integration, script or playbook mention it under the `tests:` key."
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use.py:27]:
```suggestion
    error_message = "Test playbook '{}' is not linked to any content item. Make sure at least one integration, script or playbook mentions the test-playbook id under the `tests:` key."
```

**Review by dorschw — APPROVED** [review]

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py (hunk 10 above)
Comment (shmuel44) [reply demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:329]: When all the test playbooks are in use, the query returns an empty list. I don't want the code to fail, so we have two options:
1. A list with one value - indicating that there is a test playbook not in use.
2. An empty list - indicating that all the test playbooks are in use.

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 13]
```diff
@@ -308,3 +309,28 @@ def validate_duplicate_ids(
...
+        [
+            item.get("content_items")
+            for item in run_query(tx, query)
+            if item.get("content_items")
+        ],
+        [],
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:335]:
```suggestion
        default=[],
```

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py [hunk 14]
```diff
@@ -566,6 +567,16 @@ def find_mandatory_hidden_packs_dependencies(
             self._add_relationships_to_objects(session, results)
             return [self._id_to_obj[result] for result in results]
 
+    def find_unused_test_playbook(
```
Comment (YuvHayun) [inline demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:570]: doc strings

File: demisto_sdk/commands/validate/default_config.toml [hunk 15]
```diff
@@ -34,6 +34,7 @@ select = [
   "BA116",
   "BA118",
   "DO104",
+  "GR106",
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/default_config.toml:37]: remove

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use_all_files.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    IsTestPlaybookInUseValidator,
+)
+
+ContentTypes = TestPlaybook
+
+
+class IsTestPlaybookInUseValidatorAllFiles(
+    IsTestPlaybookInUseValidator, BaseValidator[ContentTypes]
+):
+    expected_execution_mode = [ExecutionMode.SPECIFIC_FILES, ExecutionMode.USE_GIT]
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use_all_files.py:21]: I think you confused here between the class names and the expected_execution_mode

**Review by YuvHayun — COMMENTED** [review]
Looks good!
Minor comments.
Btw, no need to add any support to the old validate?
we should still support the validation there as well.

File: demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use_all_files.py (hunk 16 above)
Comment (shmuel44) [reply demisto_sdk/commands/validate/validators/GR_validators/GR106_is_testplaybook_in_use_all_files.py]: Done

File: demisto_sdk/commands/validate/default_config.toml (hunk 15 above)
Comment (shmuel44) [reply demisto_sdk/commands/validate/default_config.toml]: done

File: demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py (hunk 14 above)
Comment (shmuel44) [reply demisto_sdk/commands/content_graph/interface/neo4j/neo4j_graph.py:570]: done

**Pull request merged**